### DISCUSSION TITLE
Return a parameter specified into a kernel as output

### DISCUSSION THREAD

**Discussion by DomenicoCFD** [idaholab/moose#18801] [Q&A General] [answered]
Hi All,

I created a bespoke kernel which returns a function of space and time. To do so, I do calculate a set of intermediate parameters from my input. Basically, I would like to get returned somehow and eventually as a post-processor parameter the intermediate parameters. 

Is there a way to get that? 

Kind Regards,

Domenico

**Comment by GiudGiud** [comment]
Hello

Where is the calculation that leads to `what you want  to return` being done right now? In the input file? In the kernel?
You could move this calculation to a postprocessor (like the ParsedPostprocessor if it s simple) then feed the postprocessor in the kernel inputs.

Guillaume

**Reply by DomenicoCFD** [reply]
Hi @GiudGiud,

Many thanks for your response. There has been a bit of change of direction so that what I am trying to extract is a heat source term but now modelled as boundary condition. Having said that and following what you already suggested, I assume that the post-processor should be feed into the bc input. But how would I do that? Also, the postprocessor you suggested (i.e. ParsedPostprocessor) does not involve any object apart from other post-processors'.

I look forward to hearing from you.

Kind Regards,

**Reply by GiudGiud** [reply]
Ok so 2 questions:
How to get the calculation into the postprocessor? What are you doing to compute this term? Maybe you will need to write a new postprcoessor, maybe you can re-use existing ones

And to couple a postprocessor into a BC then you can use either of these depending on your BC:
https://mooseframework.inl.gov/source/bcs/PostprocessorDirichletBC.html
https://mooseframework.inl.gov/source/bcs/PostprocessorNeumannBC.html

sorry they are not documented. We are documenting the framework at a fast pace, this should be fixed soon

**Comment by ngrilli** [comment]
Dear @DomenicoCFD 

What I normally do is to define MaterialProperty in material objects for quantities that I would like to output.

There are many examples of this in moose, for instance:
https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/kernels/PlasticHeatEnergy.C

The plastic energy is calculated by a material object and then passed to the kernel
that applies to the temperature variable.

See the corresponding input file:
https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/jacobian/phe01.i

In that case, plastic energy can be added to an auxkernel, then visualised and postprocessed in paraview.

Nicolò

**Reply by ngrilli** [reply]
Then you use getMaterialProperty to use your quantity inside the kernel

**Comment by GiudGiud** [comment, marked as answer]
If you want the parameter to be a single value for the boundary, follow the first answer
If you want a time/space dependent value, follow the second answer.
